### PR TITLE
Refactor dependency graph

### LINKED ISSUE #1342: [FEATURE]: Support not only Notebooks in dependency graph
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

Notebooks can run files that are not notebooks, than may require migration.
The dependency graph should reflect that.

### Proposed Solution

Have the dependency graph work with any object_type, not just notebooks

### Additional Context

Sub-ticker from #1202

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
Refactor `DependencyGraph`
Introduce 'DependencyLoader' and utility classes

### Linked issues
#1202 

Resolves #1342 

### Functionality 

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests
- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

**Comment by ericvergnaud** [comment]
performance_test is throwing a cyclical-dependency error, I haven't yet found a way to get rid of it, any help appreciated...

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+
+class Dependency:
+
+    @staticmethod
+    def from_object_info(object_info: ObjectInfo):
+        assert object_info.path is not None
+        return Dependency(object_info.object_type, object_info.path)
+
+    def __init__(self, object_type: ObjectType | None, path: str):
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/dependencies.py:17]:
```suggestion
    def __init__(self, object_type: ObjectType, path: str):
```
let's simplify the handling and ignore objects with no types

File: src/databricks/labs/ucx/source_code/dependencies.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    @abc.abstractmethod
+    def object_type(self) -> ObjectType:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def build_dependency_graph(self, graph) -> None:
+        raise NotImplementedError()
+
+
+class DependencyLoader:
```
Comment (nfx) [inline src/databricks/labs/ucx/source_code/dependencies.py:48]: dependency loader has to be an interface. There has to be `LocalDependencyLoader` and `WorkspaceDependencyLoader`

**Review by nfx — CHANGES_REQUESTED** [review]
dependency loader has to be an interface, the rest is lgtm

File: src/databricks/labs/ucx/source_code/dependencies.py (hunk 2 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/source_code/dependencies.py:48]: well the intent is to manage both workspace and local dependencies from here...

**Pull request merged**